### DISCUSSION TITLE
Force use of subprojects/ dependencies

### DISCUSSION THREAD

**Discussion by kraftwerk28** [mesonbuild/meson#9734] [Q&A] [answered]
According to the [manual](https://mesonbuild.com/Dependencies.html#building-dependencies-as-subprojects):
> first Meson tries to look up the dependency from the system (such as by using pkg-config).
If it is not available, then it builds subproject

Is it possible to do the opposite, or even force to use `subprojects/` for dependency X, with only `dependency()` declaration plus some command-line arguments?
P.S. I am building a binary that specifies a [dependency](https://github.com/swaywm/sway/blob/master/meson.build#L40), which is outdated system-wide (other stable packages use it), but the most recent is in a repo, cloned inside `subprojects/`. I don't really want to tweak project's `meson.build` but instead tell meson about that behavior via command line.

**Comment by dcbaker** [comment]
You want to force the use of a fallback, even when you can find it externally?

`meson setup --force-fallback-for=dependency_name` or `meson configure --force-fallback-for=...`

should do what you want.

**Comment by eli-schwartz** [comment]
As stated by @dcbaker, meson's global option `--force-fallback-for` allows selecting subprojects to prefer. It's a core design point of meson's subproject support that no one should need to tweak `meson.build` to do this sort of thing.

However, if the dependency in this case `dependency('wayland-server', version: '>=1.20.0')` is outdated on your system and you do not *have* a suitable version, meson will normally error during configuration, because it cannot find a suitable system version.

In this case, you don't need to pass any options to use subproject dependencies, because meson automatically falls back to the subproject after system dependency lookup fails to match the version constraints.

**Comment by kraftwerk28** [comment]
Am I using it correctly?
```bash
$ rg "^\s*version:" subprojects/wayland/meson.build
3:	version: '1.20.0',

$ ls subprojects/
wayland/  wlroots/

$ meson setup --buildtype release --force-fallback-for=wayland,wayland-server build
The Meson build system
Version: 0.60.2
Source dir: /home/kraftwerk28/projects/sway/sway
Build dir: /home/kraftwerk28/projects/sway/sway/build
Build type: native build
Project name: sway
Project version: 1.6
C compiler for the host machine: cc (gcc 11.1.0 "cc (GCC) 11.1.0")
C linker for the host machine: cc ld.bfd 2.36.1
Host machine cpu family: x86_64
Host machine cpu: x86_64
Found pkg-config: /usr/bin/pkg-config (1.8.0)
Run-time dependency json-c found: YES 0.15
Run-time dependency libpcre found: YES 8.45
Dependency wayland-server found: NO found 1.19.0 but need: '>=1.20.0'
Found CMake: /usr/bin/cmake (3.22.1)
Run-time dependency wayland-server found: NO (tried cmake)

meson.build:40:0: ERROR: Invalid version of dependency, need 'wayland-server' ['>=1.20.0'] found '1.19.0'.

A full log can be found at /home/kraftwerk28/projects/sway/sway/build/meson-logs/meson-log.txt
```

Here are full logs: https://termbin.com/k640
> Line 113: Determining dependency 'wayland-server' with pkg-config executable '/usr/bin/pkg-config'

**Reply by eli-schwartz** [reply]
This verifies that meson already rejects the system version. :)

Unfortunately, in order to fallback to a subproject, one of two things needs to be true:
- the directory name (IIRC) or .wrap filename needs to match the dependency
- you need to use a wrap with a `[provide]` section

The wrap file is actually all you need, meson will download the subproject for you.

Unfortunately, meson doesn't know that "wayland" is a subproject that can provide "wayland-server".

**Reply by eli-schwartz** [reply]
I've just uploaded the dependency to the WrapDB.

Please delete your wayland subproject, then run the following command:

```
meson wrap install wayland
```

It will install the following file to subprojects/wayland.wrap:

```ini
[wrap-file]
directory = wayland-1.20.0
source_url = https://wayland.freedesktop.org/releases/wayland-1.20.0.tar.xz
source_filename = wayland-1.20.0.tar.xz
source_hash = b8a034154c7059772e0fdbd27dbfcda6c732df29cae56a82274f6ec5d7cd8725

[provide]
wayland-cursor = wayland_cursor_dep
wayland-egl = wayland_egl_dep
wayland-server = wayland_server_dep
wayland-client = wayland_client_dep
```

**Reply by kraftwerk28** [reply]
Thank you so much for the idea, I ended up writing `[wrap-git]` that looks like below:
```dosini
[wrap-git]
url = https://gitlab.freedesktop.org/wayland/wayland.git
revision = main

[provide]
dependency_names = wayland-server, wayland-client, wayland-util, wayland-private, wayland-cursor
```
and it finally compiled!

**Reply by kraftwerk28** [reply]
The question can be considered as answered

**Reply by eli-schwartz** [reply]
GitHub doesn't allow selecting a reply as an answer, apparently...

**Comment by eli-schwartz** [comment, marked as answer]
tl;dr

Use of .wrap files to map subprojects to dependency lookups is required for optimal functionality. In the case of Wayland, a stable release can be included from the WrapDB with `meson wrap install wayland`.
